### ISSUE TITLE
Editable Code Blocks

### ISSUE THREAD

**Issue by reform-chipper** [wavetermdev/waveterm#415]
I'd like to be able to edit a code block and execute it again, analogically to how Jupyter Notebooks work. 

Examples of what could this enable:
- Using workspaces as a set of steps that can be tweaked (like defining and changing all variables in a dedicated block)
- Adding markdown documentation throughout the blocks
- Keeping only those commands that ended up being useful for future reference (in the correct order)

Simple example of this vision:

<img width="1612" alt="image" src="https://github.com/wavetermdev/waveterm/assets/162800380/2af1fc2e-5e17-43e8-bb3c-89d54fbf948e">

**Comment by sawka** [comment]
Makes sense.  We'd probably have to clear the output in this case.  So it would become almost like editing "history".

I'll have to think about all of your enhancements together.  I'm not sure if a new "mode" is needed to support workbook like flows or if it is possible to hack it on to the existing in-sequence terminal model.  There are a lot of very specific things that I'd want to be able to do in "workbook" mode that don't make sense for normal terminal mode.

**Comment by reform-chipper** [comment]
I think the output doesn't have to be the cleared if you make it obvious to the user it's outdated (if the command block has been edited). 

For me it would be great to **not have a separate mode for this**. I want to be able to just use Wave to tweak and execute my commands, take some notes in between them, and have a notebook I can come back to later as a **byproduct**. 

AFAIK there really isn't something like it at the moment. You either have a dumb interactive terminal, or you have something like Runme notebooks or Jupyter with bash_kernel, where you get no interactivity at all. Thus the workflow always ends up being "work in the terminal, then remember to paste stuff that works into the static thing and hope you haven't forgotten an env variable or something". Over the years, I've developed a strong hatred for this process.

I understand designing this is a hard UI problem and I don't have all the answers. Even if it doesn't work out, Wave is already an improvement over my old way of doing things. Thanks for all the work on the project!

**Comment by carvinlo** [comment]
Just like the workflow of Warp?
<img width="1002" alt="image" src="https://github.com/user-attachments/assets/9c4e6523-291c-4f36-823d-7d33ddd8c785">

**Comment by reform-chipper** [comment]
I don't know Warp, but based on your screenshot it looks like something different. I should be able to edit any cell and run it again. Your example looks like it only works for "workflows"?

**Comment by mon-jai** [comment]
Hi @sawka. Could you explain why is this closed? Is it too hard to implement?
